src/audio/selector/selector_generic.c

```diff
@@ -171,171 +171,145 @@ static void sel_s32le_nch(struct comp_dev *dev, struct audio_stream __sparse_cac
+static void process_frame_s32le(int32_t dst[], int dst_channels,
+				int32_t src[], int src_channels,
+				struct ipc4_selector_coeffs_config *coeffs_config)
 {
-	struct audio_stream __sparse_cache *source = bsource->data;
-	struct audio_stream __sparse_cache *sink = bsink->data;
-	int8_t *src = (int8_t *)source->r_ptr;
-	int8_t *dst = (int8_t *)sink->w_ptr;
-	int bmax;
-	int b;
-	int bytes_copied = 0;
-	const int bytes_total = frames * audio_stream_frame_bytes(source);
+	int64_t accum;
+	int i;
+	int j;
 
-	while (bytes_copied < bytes_total) {
-		b = bytes_total - bytes_copied;
-		bmax = audio_stream_bytes_without_wrap(source, src);
-		b = MIN(b, bmax);
-		bmax = audio_stream_bytes_without_wrap(sink, dst);
-		b = MIN(b, bmax);
-		memcpy_s(dst, b, src, b);
-		src = audio_stream_wrap(source, src + b);
-		dst = audio_stream_wrap(sink, dst + b);
-		bytes_copied += b;
+	for (i = 0; i < dst_channels; i++) {
+		accum = 0;
+		for (j = 0; j < src_channels; j++)
+			accum += (int64_t)src[j] * (int64_t)coeffs_config->coeffs[i][j];
+		dst[i] = (int32_t)(accum >> 10);
 	}
-
-	bsource->consumed += bytes_copied >> BYTES_TO_S16_SAMPLES;
-	bsink->size += bytes_copied >> BYTES_TO_S16_SAMPLES;
 }
-#endif /* CONFIG_FORMAT_S16LE */
 
-#if CONFIG_FORMAT_S24LE || CONFIG_FORMAT_S32LE
 /**
- * \brief Channel selection for 32 bit, 1 channel data format.
- * \param[in,out] mod Selector base module device.
+ * \brief Channel selection for 32-bit, m channel input x n channel output data format.
+ * \param[in] mod Selector base module device.
  * \param[in,out] bsource Source buffer.
  * \param[in,out] bsink Sink buffer.
  * \param[in] frames Number of frames to process.
  */
-static void sel_s32le_1ch(struct processing_module *mod, struct input_stream_buffer *bsource,
-			  struct output_stream_buffer *bsink, uint32_t frames)
+static void sel_s32le(struct processing_module *mod, struct input_stream_buffer *bsource,
+		      struct output_stream_buffer *bsink, uint32_t frames)
 {
 	struct comp_data *cd = module_get_private_data(mod);
 	struct audio_stream __sparse_cache *source = bsource->data;
 	struct audio_stream __sparse_cache *sink = bsink->data;
 	int32_t *src = source->r_ptr;
-	int32_t *dest = sink->w_ptr;
-	int32_t *src_ch;
+	int32_t *dst = sink->w_ptr;
 	int nmax;
-	int i;
 	int n;
 	int processed = 0;
-	const int source_frame_bytes = audio_stream_frame_bytes(source);
-	const unsigned int nch = source->channels;
-	const unsigned int sel_channel = cd->config.sel_channel; /* 0 to nch - 1 */
+	int frm;
+	int source_frame_bytes = audio_stream_frame_bytes(source);
+	int sink_frame_bytes = audio_stream_frame_bytes(sink);
 
 	while (processed < frames) {
 		n = frames - processed;
 		nmax = audio_stream_bytes_without_wrap(source, src) / source_frame_bytes;
 		n = MIN(n, nmax);
-		nmax = audio_stream_bytes_without_wrap(sink, dest) >> BYTES_TO_S32_SAMPLES;
+		nmax = audio_stream_bytes_without_wrap(sink, dst) / sink_frame_bytes;
```

Comment:
for sel_s32le case, is sink_frame_bytes == 4 bytes or not ? Div operation will use much more cpu time

Reply:
In general, for sel_s32le sink_frame_bytes is not 4 bytes, but it is 4 bytes x sink channels count. This is because of sel_s32le being a generic function for m-channel input and n-channel output in contrast to sel_s16le_1ch (in the left panel of diff window) intended for m-channel input and 1-channel output.